### PR TITLE
Original data should not be corrupted on a delete

### PR DESCRIPTION
**Pull request by bridgetlane**
**Description**: This PR is to address the issue here: https://github.com/buger/jsonparser/issues/145

**Benchmark before change**:
goos: darwin
goarch: amd64
pkg: github.com/buger/jsonparser/benchmark
BenchmarkJsonParserLarge-8                        200000             49518 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserMedium-8                      1000000              7570 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserDeleteMedium-8                1000000              7981 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserEachKeyManualMedium-8         1000000              5192 ns/op             112 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructMedium-8         1000000              5747 ns/op             544 B/op         12 allocs/op
BenchmarkJsonParserObjectEachStructMedium-8      1000000              7763 ns/op             496 B/op         11 allocs/op
BenchmarkJsonParserSmall-8                      10000000               776 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserEachKeyManualSmall-8         20000000               599 ns/op              80 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructSmall-8         10000000               812 ns/op             192 B/op          8 allocs/op
BenchmarkJsonParserObjectEachStructSmall-8      10000000               744 ns/op             176 B/op          7 allocs/op
BenchmarkJsonParserSetSmall-8                   10000000              1097 ns/op             768 B/op          4 allocs/op
BenchmarkJsonParserDelSmall-8                    5000000              1442 ns/op               0 B/op          0 allocs/op
PASS
ok      github.com/buger/jsonparser/benchmark   104.479s

**Benchmark after change**:
goos: darwin
goarch: amd64
pkg: github.com/buger/jsonparser/benchmark
BenchmarkJsonParserLarge-8                        200000             47219 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserMedium-8                      1000000              7433 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserDeleteMedium-8                1000000              9851 ns/op           12160 B/op          4 allocs/op
BenchmarkJsonParserEachKeyManualMedium-8         1000000              5058 ns/op             112 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructMedium-8         1000000              5585 ns/op             544 B/op         12 allocs/op
BenchmarkJsonParserObjectEachStructMedium-8      1000000              7563 ns/op             496 B/op         11 allocs/op
BenchmarkJsonParserSmall-8                      10000000               796 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserEachKeyManualSmall-8         10000000               621 ns/op              80 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructSmall-8         10000000               822 ns/op             192 B/op          8 allocs/op
BenchmarkJsonParserObjectEachStructSmall-8      10000000               730 ns/op             176 B/op          7 allocs/op
BenchmarkJsonParserSetSmall-8                   10000000              1069 ns/op             768 B/op          4 allocs/op
BenchmarkJsonParserDelSmall-8                    5000000              1704 ns/op             720 B/op          4 allocs/op
PASS
ok      github.com/buger/jsonparser/benchmark   101.300s

**Time comparison using benchstat**
name      old time/op    new time/op    delta
Delete-8    20.6µs ± 9%    22.7µs ± 4%  +10.04%  (p=0.000 n=10+9)

name      old alloc/op   new alloc/op   delta
Delete-8    16.5kB ± 0%    32.9kB ± 0%  +99.13%  (p=0.000 n=10+10)

name      old allocs/op  new allocs/op  delta
Delete-8      2.00 ± 0%      3.00 ± 0%  +50.00%  (p=0.000 n=10+10)

For running benchmarks use:
```
go test -test.benchmem -bench JsonParser ./benchmark/ -benchtime 5s -v
# OR
make bench (runs inside docker)
```

### CODE REVIEW

File: parser.go [hunk 1]
```diff
@@ -704,7 +704,12 @@ func Delete(data []byte, keys ...string) []byte {
 		newOffset = prevTok + 1
 	}
 
-	data = append(data[:newOffset], data[endOffset:]...)
+	// We have to make a copy here if we don't want to mangle the original data, because byte arrays are
```
Comment (tkuhlman) [inline parser.go]: `byte slices are accessed by reference` it is a slice not an array which is pretty key distinction regarding reference/value passing.

File: parser_test.go [hunk 2]
```diff
@@ -1074,15 +1074,23 @@ func runSetTests(t *testing.T, testKind string, tests []SetTest, runner func(Set
...
+		original := make([]byte, len(test.json))
+		copy(original, test.json)
+
 		fmt.Println("Running:", test.desc)
 
-		value := runner(test)
+		value, bytes := runner(test)
+
+		if string(original) != string(bytes) {
+			t.Errorf("ORIGINAL DATA MALFORMED: %v, %v", string(original), string(bytes))
```
Comment (tkuhlman) [inline parser_test.go:1091]: Good job remembering to add a test to catch this.

**Review by tkuhlman — APPROVED** [review]
Code looks great, one minor point regarding the comment is all I saw.

File: parser.go (hunk 1 above)
Comment (bridgetlane) [reply parser.go]: whoops, thanks for the catch. I'll fix

**Pull request merged**